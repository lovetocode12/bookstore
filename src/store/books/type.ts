
export const GET_BOOKS = 'GET_BOOKS'
export const GET_BOOK_INPROGRESS = 'GET_BOOK_INPROGRESS'
export const GET_BOOK_SUCCESS = 'GET_BOOK_SUCCESS'
export const GET_BOOK_FAILED = 'GET_BOOK_FAILED'

export interface Book {
  id: string,
  title: string,
  description: string,
  url: string,
  price: number,
  pageCount: number,
  author: string
}



export interface BookState {
  loading: boolean;
  books: Book[];
  error: string;
}

interface GETBooksAction {
  type: typeof GET_BOOKS;
}

export interface GetBookInProgressAction {
  type: typeof GET_BOOK_INPROGRESS;
}

export interface GetBookSuccessAction {
  type: typeof GET_BOOK_SUCCESS;
  payload: Book[]
}

export interface GetBookFailedAction {
  type: typeof GET_BOOK_FAILED;
  payload: string
}

export type BooksActionTypes = GETBooksAction  
| GetBookInProgressAction | GetBookSuccessAction | GetBookFailedAction;

