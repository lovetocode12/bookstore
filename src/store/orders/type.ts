import { Book } from "../books/type";

export const ADDTOCART_BOOK = 'ADDTOCART_BOOK'
export const BUYNOW_BOOK = 'BUYNOW_BOOK'
export const CHECKOUT_CART = 'CHECKOUT_CART'
export const CLEAR_CART = 'CLEAR_CART'
export const GET_ADDRESS = 'GET_ADDRESS'
export const SET_ADDRESS = 'SET_ADDRESS'

export interface Order {
    book: Book
    orderStatus: 'InProcess' | 'Delivered'
    orderDate: Date
}

export interface OrderState {
        pastOrders: Array<Order>,
        currentOrders: Array<Order>
        address: Address
}


export interface Address {
    address1: string;
    address2: string;
    email: string;
    phone: string;
    city: string;
}

export interface AddToCartBOOKAction {
    type: typeof ADDTOCART_BOOK;
    payload: Order;
}

export interface BuyNowBooksAction {
    type: typeof BUYNOW_BOOK;
    payload: Order
}

export interface ClearCartAction {
    type: typeof CLEAR_CART;
}

export interface CheckoutCartAction {
    type: typeof CHECKOUT_CART;
}

export interface SetAddressAction {
    type: typeof SET_ADDRESS;
    payload: Address
}

export interface GetAddressAction {
    type: typeof GET_ADDRESS;
}

export type OrdersActionTypes = AddToCartBOOKAction | BuyNowBooksAction 
| ClearCartAction | CheckoutCartAction | GetAddressAction | SetAddressAction;

