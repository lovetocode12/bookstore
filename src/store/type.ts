import { BookState } from "./books/type";
import { OrderState } from "./orders/type";


export interface RootState {
    bookState: BookState;
    orderState: OrderState;
}