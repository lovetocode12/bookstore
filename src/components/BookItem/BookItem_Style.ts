import styled from "styled-components";

export const CardDiv = styled.div`
    display: flex;
    flex-wrap: wrap;
    flex-basis: 12em;
    padding: 5px;
    border-radius: 5px;
    min-height: 220px;
    justify-content: center;
    align-items: baseline;
    flex-grow: 1;
    flex-shrink: 1;
    margin: 1em 1em;
    box-shadow: 2px 2px 3px #eeeeee9e;;
    `

    export const PhotoDiv = styled.div`
    display: flex;
    width: 80%;
    justify-content: center;
    cursor: pointer;
    background-color: #e2c7c73b;
    `

    export const Image = styled.img`
    width: 60%;
    height: 100%;
    border-radius: 2em;
    object-fit: cover;
    padding: 1em;
    `

    export const ContentDiv = styled.div`
    display: flex;
    flex-wrap: wrap;
    justify-content: space-between;
    align-items: center;`

    export const PTag = styled.p`
    width: 12rem;
    font-size: 0.8em;
    overflow: hidden;
    white-space: nowrap;
    display: inline-block;
    text-overflow: ellipsis;
    margin: 0;`

    export const H4Tag = styled.h4`
    width: 12rem;
    overflow: hidden;
    white-space: nowrap;
    display: inline-block;
    text-overflow: ellipsis;
    margin: 0;
    font-size: 0.8em;
    text-align: center;`

    export const ButtonsContainer = styled.div`
    display: flex;
    justify-content: space-evenly;
    align-items: center;
    width: 80%;
    flex-wrap: wrap;
    margin-top: 0.4em;
`

    export const StyledButton = styled.button`
    margin: 0;
    color: white;
    background: #ea8686;
    padding: 0.5em;
    font-size: 0.7rem;
    cursor: pointer;
    border-radius: 5px;
    border: none;`
