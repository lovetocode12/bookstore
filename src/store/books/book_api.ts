export function GetBooksApi(){
  return  fetch('http://localhost:3001/books')
    .then(res=> res.json())
    .then(books=> books)
}