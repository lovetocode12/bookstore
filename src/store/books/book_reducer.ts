import { BookState, BooksActionTypes, GET_BOOK_INPROGRESS, GET_BOOK_SUCCESS, GET_BOOK_FAILED } from "./type";

export const initialBookState: BookState = {
    loading: false,
    books: [],
    error: ''
};

export function bookReducer(state: BookState = initialBookState , action: BooksActionTypes): BookState {
    switch (action.type) {
        case GET_BOOK_INPROGRESS:
            return {
                ...state, loading: true
            };
        case GET_BOOK_SUCCESS:
            return {
                ...state,
                loading: false,
                books:[...state.books,...action.payload]
            };
        case GET_BOOK_FAILED:
            return {
                ...state,
                loading: false,
                error:action.payload
            };
        default:
            return state;
    }
}