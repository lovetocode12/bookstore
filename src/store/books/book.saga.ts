import { takeEvery, call, put } from "redux-saga/effects"
import { GetBooksApi } from "./book_api"
import { GET_BOOKS } from "./type";
import { getBooksInProgress, getBooksSuccess, getBooksFailed } from "./books_action";

export function* getBooksSaga() {
    yield put(getBooksInProgress());
    try {
        const books = yield call(GetBooksApi);
        yield put(getBooksSuccess(books));
    } catch (error) {
        yield put(getBooksFailed(error));
    }

}

export function* WatchBooks() {
    yield takeEvery(GET_BOOKS, getBooksSaga)
}