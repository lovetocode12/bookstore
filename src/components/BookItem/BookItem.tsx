import React, { SyntheticEvent } from 'react'
import { Book } from '../../store/books/type'
import {
    CardDiv, PhotoDiv, Image, ContentDiv,
    H4Tag, PTag, StyledButton, ButtonsContainer
} from './BookItem_Style'
import { useHistory } from 'react-router-dom'
import { Order } from '../../store/orders/type'
import useBuyBook from '../../common/useBuyBook'

type Props = Book

const BookItem: React.FC<Props> = ({ id, title, url, description, price, pageCount, author }) => {

    const history = useHistory();
    const [BuyNow, AddToCart] = useBuyBook();

    function NavToDetailsPage(event: SyntheticEvent) {
        history.push(`/details/${id}`)
        event.stopPropagation();
    }

    function BuyNowClick() {
        let order: Order = {
            orderStatus: 'InProcess',
            orderDate: new Date(),
            book: {
                id, description,
                url, title, price, author, pageCount
            }

        }
        BuyNow(order);
    }

    function AddToCartClick() {
        let order: Order = {
            orderStatus: 'InProcess',
            orderDate: new Date(),
            book: {
                id, description,
                url, title, price, author, pageCount
            }

        }
        AddToCart(order);
    }

    return (
        <CardDiv>
            <PhotoDiv onClick={(e) => NavToDetailsPage(e)}>
                <Image src={url} alt={title} ></Image>
            </PhotoDiv>
            <ContentDiv>
                <H4Tag>{title}</H4Tag>
                <PTag> {description} </PTag>
                <ButtonsContainer>
                    <StyledButton onClick={BuyNowClick}>Buy Now</StyledButton>
                    <StyledButton onClick={AddToCartClick}>Add To Cart</StyledButton>
                </ButtonsContainer>

            </ContentDiv>
        </CardDiv>
    )
}

export default React.memo(BookItem)
