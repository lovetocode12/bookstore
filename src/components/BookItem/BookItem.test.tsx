import React from 'react';
import { shallow } from '../../../enzyme';
import { Book } from '../../store/books/type';
import BookItem from './BookItem';
import { H4Tag, StyledButton } from './BookItem_Style';

const mockDispatch = jest.fn();
jest.mock('react-redux', () => ({
    useDispatch: () => mockDispatch
}));

describe('BookList Component', () => {

    const book = {
        title: "Flex on Java",
        id: "1933988797",
        pageCount: 265,
        price: 22,
        url: "https://s3.amazonaws.com/AKIAJC5RLADLUMVRPFDQ.book-thumb-images/allmon.jpg",
        description: "Software Solutions Ltd.",
        author: "Bernerd Allmon"
    };
    function getWrapper() {
        const props: Book = book;

        const wrapper: any = shallow(<BookItem {...props} />);
        return wrapper;
    }

    it('Component should render correctly', () => {
        const wrapper = getWrapper();
        expect(wrapper).toMatchSnapshot();
    })

    it('Should display title as Flex on Java', () => {
        const wrapper = getWrapper();
        expect(wrapper.find(H4Tag).text()).toContain('Flex on Java');
    })

    it('Should call dispatch with type BUYNOW_BOOK', () => {
        const wrapper = getWrapper();
        const buyNowButton = wrapper.find(StyledButton);
        buyNowButton.at(1).simulate('click');
        expect(mockDispatch).toHaveBeenCalledTimes(1);
    })
});