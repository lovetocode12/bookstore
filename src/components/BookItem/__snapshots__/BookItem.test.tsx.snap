// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookList Component Component should render correctly 1`] = `
<styled.div>
  <styled.div
    onClick={[Function]}
  >
    <styled.img
      alt="Flex on Java"
      src="https://s3.amazonaws.com/AKIAJC5RLADLUMVRPFDQ.book-thumb-images/allmon.jpg"
    />
  </styled.div>
  <styled.div>
    <styled.h4>
      Flex on Java
    </styled.h4>
    <styled.p>
       
      Software Solutions Ltd.
       
    </styled.p>
    <styled.div>
      <styled.button
        onClick={[Function]}
      >
        Buy Now
      </styled.button>
      <styled.button
        onClick={[Function]}
      >
        Add To Cart
      </styled.button>
    </styled.div>
  </styled.div>
</styled.div>
`;
